File: proposed-projects/round-trip-experiment/reverse/20260909T180307Z_280f3890/20260910T095157Z_2fde87b4/ported_implementation/gbnf/grammar_graph/pointers.py
from typing import Dict, Iterator


class Pointers:
    """An insertion-ordered collection of pointers, keyed by pointer id.

    A pointer's id is the sum of its node's id and its parent's id chain, so two
    pointers sharing an id point at the same node through an identical parent
    chain and are interchangeable when walking the graph.
    """

    def __init__(self, *pointers):
        self._pointers: Dict[str, object] = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer) -> None:
        self._pointers[pointer.id] = pointer

    def __iter__(self) -> Iterator:
        yield from list(self._pointers.values())

    @property
    def size(self) -> int:
        return len(self._pointers)

File: packages/gbnf-experiment/data/20260909T180307Z_280f3890/ported_implementation/src/grammar-graph/pointers.ts
import type { GraphPointer } from './graph-pointer.js';

/**
 * An insertion-ordered collection of pointers, keyed by pointer id.
 *
 * A pointer's id is the sum of its node's id and its parent's id chain, so two
 * pointers sharing an id point at the same node through an identical parent
 * chain and are interchangeable when walking the graph.
 */
export class Pointers {
  readonly #pointers = new Map<string, GraphPointer>();

  constructor(...pointers: GraphPointer[]) {
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.#pointers.set(pointer.id, pointer);
  }

  *[Symbol.iterator](): Generator<GraphPointer> {
    yield* this.#pointers.values();
  }

  get size(): number {
    return this.#pointers.size;
  }
}
